sci-libs/opencascade/opencascade-7.5.2.ebuild

```diff
@@ -0,0 +1,175 @@
+	"${FILESDIR}"/${PN}-7.5.1-fix-AllValues-name-collision-with-vtk-9.0.patch
+)
+
+src_prepare() {
+	cmake_src_prepare
+
+	if use debug; then
+		append-cppflags -DDEBUG
+		append-flags -g
+	fi
+
+	sed -e 's/\/lib\$/\/'$(get_libdir)'\$/' \
+		-i adm/templates/OpenCASCADEConfig.cmake.in || die
+
+	# There is an OCCT_UPDATE_TARGET_FILE cmake macro that fails due to some
+	# assumptions it makes about installation paths. Rather than fixing it, just
+	# get rid of the mechanism altogether - its purpose is to allow a
+	# side-by-side installation of release and debug libraries.
+	sed -e 's|\\${OCCT_INSTALL_BIN_LETTER}||' \
+		-i "adm/cmake/occt_toolkit.cmake" || die
+}
+
+src_configure() {
+	local mycmakeargs=(
+		-DBUILD_DOC_Overview=$(usex doc)
+		-DBUILD_Inspector=$(usex examples)
+		-DBUILD_WITH_DEBUG=$(usex debug)
+		-DCMAKE_INSTALL_PREFIX="/usr"
+		-DINSTALL_DIR_BIN="$(get_libdir)/${P}/bin"
+		-DINSTALL_DIR_CMAKE="$(get_libdir)/cmake/${P}"
+		-DINSTALL_DIR_DOC="share/doc/${PF}"
+		-DINSTALL_DIR_LIB="$(get_libdir)/${P}"
+		-DINSTALL_DIR_SCRIPT="$(get_libdir)/${P}/bin"
+		-DINSTALL_DIR_WITH_VERSION=ON
+		-DINSTALL_SAMPLES=$(usex examples)
+		-DINSTALL_TEST_CASES=NO
+		-DUSE_D3D=NO
+		-DUSE_FFMPEG=$(usex ffmpeg)
+		-DUSE_FREEIMAGE=$(usex freeimage)
+		-DUSE_FREETYPE=ON
+		-DUSE_GLES2=$(usex gles2)
+		-DUSE_RAPIDJSON=$(usex json)
+		-DUSE_TBB=$(usex tbb)
+		-DUSE_VTK=$(usex vtk)
+	)
+
+	use doc && mycmakeargs+=( -DINSTALL_DOC_Overview=ON )
+
+	if use examples; then
+		mycmakeargs+=(
+			-D3RDPARTY_QT_DIR="${ESYSROOT}"/usr
+			-DBUILD_SAMPLES_QT=ON
+		)
+	fi
+
+	if use vtk; then
+		if has_version ">=sci-libs/vtk-9.0.0"; then
+			mycmakeargs+=(
+				-D3RDPARTY_VTK_DIR="${ESYSROOT}"/usr
+				-D3RDPARTY_VTK_INCLUDE_DIR="${ESYSROOT}"/usr/include/vtk-9.0
+				-D3RDPARTY_VTK_LIBRARY_DIR="${ESYSROOT}"/usr/$(get_libdir)
+			)
+		fi
+	fi
+
+	cmake_src_configure
+
+	prepare_env_file() {
+		# prepare /etc/env.d file
+		sed -e 's|VAR_CASROOT|'${ESYSROOT}'/usr|g' < "${FILESDIR}/${PN}-${PV_MAJ}.env.in" >> "${T}/${PV_MAJ}" || die
+		sed -e 's|lib/|'$(get_libdir)'/|g' -i "${T}/${PV_MAJ}" || die
+		sed -e 's|VAR_PV|'${PV}'|g' -i "${T}/${PV_MAJ}" || die
+
+		# use TBB for memory allocation optimizations?
+		use tbb && (sed -e 's|^#MMGT_OPT=0$|MMGT_OPT=2|' -i "${T}/${PV_MAJ}" || die)
+
+		if use optimize ; then
+			# use internal optimized memory manager?
+			sed -e 's|^#MMGT_OPT=0$|MMGT_OPT=1|' -i "${T}/${PV_MAJ}" || die
+			# don't clear memory ?
+			sed -e 's|^#MMGT_CLEAR=1$|MMGT_CLEAR=0|' -i "${T}/${PV_MAJ}" || die
+		fi
+	}
+
+	prepare_env_file
+}
+
+src_install() {
+	use doc && docompress -x /usr/share/doc/${PF}/overview/html
+	cmake_src_install
+
+	# respect slotting
+	insinto "/etc/env.d/${PN}"
+	doins "${T}/${PV_MAJ}"
+
+	# remove examples
+	if use !examples; then
+		rm -r "${ED}/usr/share/${P}/samples" || die
+	fi
+}
+
+pkg_postinst() {
+	eselect ${PN} set ${PV_MAJ} || die "failed to switch to updated implementation"
```

Comment:
The eselect doesn't have an update function, it's very basic. TARGET and CONFDIR dir are ${EROOT} aware.

I'm no longer sure, I should use this eselect here. When I introduced it, I was thinking to have it updated, so a possible rebuild of freecad can pick it up. But meanwhile I have got reportings of users, where freecad fails to rebuild because of this, because the environment changes didn't get re-sourced during an emerge, but only at the end of it.
So I think it would be better to remove this line. This however could require people to rebuild freecad two times, the first one triggered by portage, the second time, after the eselect module and the environment are updated.
Any hints on how cleanly solve this?

Reply:
If possible it'd be nice if freecad was able to select the implementation itself, much like how python ebuilds pick the interpreter disregarding what the user have eselect'ed.

Reply:
On a side-note, other ebuilds seem to not agree on how to use this.

sci-electronics/kicad will likely break with 7.5.2, and dealii explicitly requests 7.5.1 as a workaround.

Reply:
> If possible it'd be nice if freecad was able to select the implementation itself, much like how python ebuilds pick the interpreter disregarding what the user have eselect'ed.

It did, but freecad upstream supports any v7 release of occ, maybe even v6.4 is still supported, not sure about the last. But meanwhile the last revdep has been fixed to work with 7.5.2 as well. The issue with the failure was, because I changed the installation location of opencascade starting with 7.5. Earlier versions, including 7.4 installed in some obscure `/usr/$(get_libdir)/${P}/ros/` directory. So an update from 7.4 to 7.5 caused a major change in the environment variables set by the package, which caused the build failure, because the environment used by emerge still hold the old location. I would guess, other packages would also fail, if you try to rebuild them after updating 7.4 to 7.5 before the new environment is sourced. I think this was a one time issue, as newer versions will also use the current layout.

Reply:
> sci-electronics/kicad will likely break with 7.5.2 (it hardcodes -7.5.1 paths if has_version >=7.5.1), and dealii explicitly requests 7.5.1 as a workaround (so this one won't break and can take time to fix this, just means not being able to cleanup old for a while which makes sense not to anyway).

Yes I see. Many packages still don't use cmake config mode to get the correct information and set the obsolete `OCC_{INCLUDE,LIBRARY}_DIR` variables manually. This includes kicad and freecad. With freecad I use a way which doesn't hardcode the version with the include and library dir, but might be more volatile to changes. Kicad could eventually use the same approach. I'm planning on re-working the parts for opencascade in the freecad ebuilds, once v7.4 has been removed from the tree.

Reply:
> **Edit2:** also, how does this behave if different ROOT? Just fail or is it ${EROOT} aware? Aka, to update in a chroot target or when cross-compiling.

Don't really know. TBH I copied most of the code from other eselect modules and adapted it. Didn't spend too much time with the details of the eselect modules yet. The module could probably need some attention. But so far it worked, and I have more important issues at hand which I focus priority on.

Reply:
Have to ask, any chances of simply unslotting this? Debian for instance just use /usr/include/opencascade for 7.5.1 and I imagine it's a lot less problems. For us it'd mean no eselect-opencascade, no env file to get paths, no ebuild workarounds, etc... with just a normal subslot for rebuilds.

In most case were try to bring packages up to par so we can just use latest versions rather than keep multiple.

I'd just like this to be in a better state rather than keep moving forward with a broken setup.

Reply:
> I'd just like this to be in a better state rather than keep moving forward with a broken setup.

Well, it could be left for next time, but please give it some thoughts.

Reply:
> Have to ask, any chances of simply unslotting this? Debian for instance just use /usr/include/opencascade for 7.5.1 and I imagine it's a lot less problems. For us it'd mean no eselect-opencascade, no env file to get paths, no ebuild workarounds, etc... with just a normal subslot for rebuilds.
> 

I was thinking about this several times myself. Need to see, if I can recollect the details of why we introduced the eselect module in the first place. It had something to do with the switch from v6.x to v7.x.
It wouldn't mean to get rid of the environment file. It will still be needed, as the package suspects these variables to be available in the environment at runtime, but it will nonetheless make the ebuild simpler and we can get rid of the eselect module. I keep this in mind for the next bump.

> In most case we try to bring packages up to par so we can just use latest versions rather than keep multiple.
> **Edit:** but nevermind if that's really needed -- debian for instance can more easily build things against a different version, just want to know what you think about it.
> 
> I'd just like this to be in a better state rather than keep moving forward with a broken setup.

The issues with the broken setup has a multitude of reasons which are tied to interdependant packages (vtk-9, opencascade and freecad) being merged in a different order than I had them pushed and no easy way to test them against  each other with the new, not yet merged versions and not being compatible upstream without patches. The setup wasn't broken before vtk-9 and occ-7.5 came up, and I suspect them to be fixed again soonish, with the removal of obsolete occ-7.4 and stabilization and wider adoption of vtk-9.

Reply:
Thanks for your thoughts and info.

And yeah env file for runtime is no issue, I was more thinking in the context of build time preferably not needing those values if possible.

Anyhow, let's keep this for now -- been testing 7.5.2 and freecad a bit and things are looking fine from my end.